### ISSUE TITLE
TCP/UDP: System.Net.Sockets.SocketException : An address incompatible with the requested protocol was used

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#4071]
http://dotnet-ci.cloudapp.net/job/dotnet_corefx_prtest/5713/

```
System.Net.Sockets.Tests.DualMode.BeginSendToV4IPEndPointToDualHost_Success [FAIL]
        System.Net.Sockets.SocketException : An address incompatible with the requested protocol was used
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\src\System\Net\Sockets\Socket.cs(3210,0): at System.Net.Sockets.Socket.DoBeginSendTo(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint endPointSnapshot, SocketAddress socketAddress, OverlappedAsyncResult asyncResult)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\src\System\Net\Sockets\Socket.cs(3161,0): at System.Net.Sockets.Socket.BeginSendTo(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint remoteEP, AsyncCallback callback, Object state)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\DualModeSocketTest.cs(1364,0): at System.Net.Sockets.Tests.DualMode.DualModeBeginSendTo_EndPointToHost_Helper(IPAddress connectTo, IPAddress listenOn, Boolean dualModeServer, Boolean expectedToTimeout)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\DualModeSocketTest.cs(1350,0): at System.Net.Sockets.Tests.DualMode.BeginSendToV4IPEndPointToDualHost_Success()
```

cc: @cipop, @pgavlin

**Comment by CIPop** [comment]
The stack is out of sync with the code in master. 

We need to be able to repro or to collect a full memory dump to investigate the state of the socket at the time of the failure.

Please reopen the issue if it's seen again in the CI system.

**Comment by stephentoub** [comment]
@CIPop, it just repro'd again:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_release_prtest/6212/console
```
System.Net.Sockets.Tests.DualMode.BeginSendToV4IPEndPointToDualHost_Success [FAIL]
        System.Net.Sockets.SocketException : An address incompatible with the requested protocol was used
        Stack Trace:
  Starting:    System.Net.Primitives.Functional.Tests
  Discovering: System.Security.Cryptography.Csp.Tests
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\src\System\Net\Sockets\Socket.cs(2941,0): at System.Net.Sockets.Socket.DoBeginSendTo(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint endPointSnapshot, SocketAddress socketAddress, OverlappedAsyncResult asyncResult)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\src\System\Net\Sockets\Socket.cs(2892,0): at System.Net.Sockets.Socket.BeginSendTo(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint remoteEP, AsyncCallback callback, Object state)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\src\System\Net\Sockets\SocketTaskExtensions.cs(234,0): at System.Net.Sockets.SocketTaskExtensions.<SendToAsync>b__32(ArraySegment`1 targetBuffer, SocketFlags flags, EndPoint endPoint, AsyncCallback callback, Object state)
              at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\src\System\Net\Sockets\SocketTaskExtensions.cs(233,0): at System.Net.Sockets.SocketTaskExtensions.SendToAsync(Socket socket, ArraySegment`1 buffer, SocketFlags socketFlags, EndPoint remoteEndPoint)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\DualModeSocketTest.cs(1392,0): at System.Net.Sockets.Tests.DualMode.DualModeBeginSendTo_EndPointToHost_Helper(IPAddress connectTo, IPAddress listenOn, Boolean dualModeServer, Boolean expectedToTimeout)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets.Legacy\tests\FunctionalTests\DualModeSocketTest.cs(1378,0): at System.Net.Sockets.Tests.DualMode.BeginSendToV4IPEndPointToDualHost_Success()
```

**Comment by SidharthNabar** [comment]
Need to investigate this test - this seems to be a test setup/infra issue, hence moving to RTM milestone.

**Comment by stephentoub** [comment]
Failed again here: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/windows_nt_release_prtest/917/

**Comment by CIPop** [comment]
According to @josguil this could be a code bug. Keeping this as the only tracking issue.

> I don't think it's a test issue. The error message happens when the address family of client and server are different. But both server and client are initialized to IPv4 in the test. The one for the UdpClient is by default Ipv4 even if it's not passed explicitly to the constructor.
> 
> It looks like something else could be changing one of the addresses, my best bet is CheckCacheRemote method, since it can actually > remap an Ipv4 endpoint to be Ipv6. But I couldn't get a repro, and I need it to give a better answer on what path our code is following in > this method when it throws the exception.

**Comment by CIPop** [comment]
I don't have a repro for this yet.

**Comment by steveharter** [comment]
Keeping this as an open issue. Last repro on 1/13/2017. However this occurred on mac (not Windows):
```
dotnet_corefx/master/osx_release 621 	dci-macpro-09 	01/13 11:43 PM
```

**Comment by stephentoub** [comment]
@geoffkizer, you were recently telling me about a case where a socket address object should have been pinned but wasn't, right?  Was that on this code path?  I'm wondering if that was the cause of this, that we passed the address of an unpinned array to native, the object got moved and stomped on, and then Windows read bad data, resulting in it saying the address wasn't valid.

**Comment by geoffkizer** [comment]
#16862 is the PR for the SendTo fix.

**Comment by stephentoub** [comment]
I'm going to close this under the assumption that it's fixed by #16862.  We can reopen if it turns out that wasn't the cause/fix and it happens again.
